src/snowflake/snowpark/_internal/analyzer/common_subdataframe_elimination.py

```diff
@@ -0,0 +1,209 @@
+                with_block = WithQueryBlock(
+                    name=random_name_for_temp_object(TempObjectType.CTE),
+                    child=node
+                )
+                # resolve with block
+                resolved_with_block = self._analyzer.do_resolve_with_resolved_children(with_block, resolved_nodes, df_aliased_col_name_to_real_col_name)
+                resolved_nodes[with_block] = resolved_with_block
+                for parent in parents:
+                    if isinstance(parent, SelectStatement):
+                        parent._from = WithObjectRef(with_block)
+                        parent._snowflake_plan = resolved_with_block
+                    elif isinstance(parent, SetStatement):
+                        parent._nodes.clear()
+                        for operand in parent.set_operands:
+                            if operand.selectable == node:
+                                with_object_ref_node = WithObjectRef(with_block)
+                                with_object_plan = self._analyzer.do_resolve_with_resolved_children(
+                                    with_object_ref_node,
+                                    resolved_nodes,
+                                    resolved_with_block.df_aliased_col_name_to_real_col_name
+                                )
+                                new_selectable = SelectSnowflakePlan(
+                                    snowflake_plan=with_object_plan,
+                                    analyzer=self._analyzer
+                                )
+                                operand.selectable = new_selectable
+                                parent._nodes.append(new_selectable)
+                            else:
+                                parent._nodes.append(operand.selectable)
+
+                        # refresh the sql query
+                        parent._sql_query = None
+                        parent._snowflake_plan = None
+
+                    elif isinstance(parent, SnowflakePlan):
+                        for i in range(len(parent.source_plan.children)):
+                            try:
+                                if parent.source_plan.children[i] == node:
+                                    parent.source_plan.children[i] = WithObjectRef(with_block)
+                            except Exception as e:
+                                pass
+
+            else:
+                if isinstance(node, SnowflakePlan):
+                    res = self._analyzer.do_resolve_with_resolved_children(
+                        node.source_plan,
+                        resolved_nodes,
+                        df_aliased_col_name_to_real_col_name
+                    )
+                    node.schema_query = res.schema_query
+                    node.queries = res.queries
+                    node.with_query_block_plans = res.with_query_block_plans
+                    resolved_nodes[node] = node
```

Comment:
why not  `resolved_nodes[node] = res`? This is what we do when `node` is `Selectable`